### PR TITLE
fix: benchmark storage issue

### PR DESCRIPTION
**Pull request by weichweich**
* The storage seems to be dirty at the start of a benchmark

## Checklist:

- [ ] I have verified that the code works
  - [ ] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] This PR does not introduce new custom types
  - [ ] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: pallets/delegation/src/tests.rs [hunk 1]
```diff
@@ -136,17 +136,17 @@ fn create_delegation_direct_root_successful() {
...
-				generate_base_delegation_node(hierarchy_root_id, delegate, Some(hierarchy_root_id)),
+				generate_base_delegation_node(hierarchy_root_id, delegate.clone(), Some(hierarchy_root_id)),
 			);
 			let delegation_info = Delegation::calculate_delegation_creation_hash(
 				&delegation_id,
 				&hierarchy_root_id,
 				&hierarchy_root_id,
 				&delegation_node.details.permissions,
 			);
-			let delegate_signature = delegate_keypair.sign(&hash_to_u8(delegation_info));
```
Comment (wischli) [inline pallets/delegation/src/tests.rs:147]: S: Seems like we don't need the `delegate_keypair` variable at all anymore, so we can just get rid of it?

File: support/src/signature.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		Weight::zero()
+	}
+}
+
+#[cfg(any(test, feature = "mock", feature = "runtime-benchmarks"))]
+pub struct EqualVerify<A, B>(PhantomData<(A, B)>);
+#[cfg(any(test, feature = "mock", feature = "runtime-benchmarks"))]
+impl<Account, Payload> VerifyDelegateSignature for EqualVerify<Account, Payload>
+where
+	Account: PartialEq + Clone,
```
Comment (Diiaablo95) [inline support/src/signature.rs:86]: Why do we need `Clone` here?

File: runtimes/standalone/Cargo.toml [hunk 3]
```diff
@@ -76,6 +77,9 @@ runtime-benchmarks = [
   "frame-system-benchmarking",
   "hex-literal",
   "kilt-launch/runtime-benchmarks",
+  "kilt-primitives/runtime-benchmarks",
+  "kilt-support",
```
Comment (Diiaablo95) [inline runtimes/standalone/Cargo.toml:81]: I think we only use `AlwaysVerify` from `kilt-support`, so it should work also removing this?

File: runtimes/spiritnet/Cargo.toml [hunk 4]
```diff
@@ -93,6 +94,8 @@ runtime-benchmarks = [
   "frame-system-benchmarking",
   "hex-literal",
   "kilt-launch/runtime-benchmarks",
+  "kilt-support",
```
Comment (Diiaablo95) [inline runtimes/spiritnet/Cargo.toml:97]: Same here perhaps? I mean here we don't even use the crate yet.

File: runtimes/peregrine/Cargo.toml [hunk 5]
```diff
@@ -100,6 +101,9 @@ runtime-benchmarks = [
   "frame-system-benchmarking",
   "frame-system/runtime-benchmarks",
   "hex-literal",
+  "kilt-primitives/runtime-benchmarks",
+  "kilt-support",
```
Comment (Diiaablo95) [inline runtimes/peregrine/Cargo.toml:105]: Same here?

File: pallets/did/src/benchmarking.rs [hunk 6]
```diff
@@ -101,7 +104,11 @@ fn generate_base_did_call_operation<T: Config>(
 // encoded and which key takes the longest to verify and always use that.
```
Comment (Diiaablo95) [inline pallets/did/src/benchmarking.rs:104]: I would remove this comment by now then. We have found a workaround to that.

File: pallets/did/src/benchmarking.rs [hunk 7]
```diff
@@ -759,6 +766,88 @@ benchmarks! {
 	verify {
 		assert!(!Did::<T>::get(&did_subject).unwrap().key_agreement_keys.contains(&key_agreement_key_id));
 	}
+
+	signature_verification_sr {
```
Comment (Diiaablo95) [inline pallets/did/src/benchmarking.rs:770]: C: I would add the final `25519`, just in case.

File: pallets/did/src/benchmarking.rs [hunk 8]
```diff
@@ -759,6 +766,88 @@ benchmarks! {
...
+		assert_ok!(did_details.update_attestation_key(DidVerificationKey::from(get_ecdsa_public_attestation_key()), block_number));
+
+		Did::<T>::insert(&did_subject, did_details);
+		let signature = sr25519_sign(AUTHENTICATION_KEY_ID, &public_auth_key, &payload).expect("Failed to create DID signature from raw sr25519 signature.");
+		let did_signature = DidSignature::Sr25519(signature);
+	}: {
+		DidSignatureVerify::<T>::verify(&did_subject, &payload, &did_signature).expect("should verify");
+	}
+	verify {}
+	signature_verification_ed {
```
Comment (Diiaablo95) [inline pallets/did/src/benchmarking.rs:797]: Also here...

File: pallets/did/src/benchmarking.rs [hunk 9]
```diff
@@ -37,6 +39,7 @@ const AUTHENTICATION_KEY_ID: KeyTypeId = KeyTypeId(*b"0000");
 const ATTESTATION_KEY_ID: KeyTypeId = KeyTypeId(*b"0001");
 const DELEGATION_KEY_ID: KeyTypeId = KeyTypeId(*b"0002");
 const UNUSED_KEY_ID: KeyTypeId = KeyTypeId(*b"1111");
+const MAX_SIGNATURE_BYTE_LENGTH: u32 = 5 * 1024 * 1024;
```
Comment (Diiaablo95) [inline pallets/did/src/benchmarking.rs:42]: Maybe not part of this PR, but should we make this part of the pallet `Config` then? Or, alternatively, should this value be the maximum between `Ed25519`, `Sr25519`, and `Ecdsa` signature lenghts? That would be more realistic, as the longest signature is 64 bytes and here we are definitely overestimating.

File: pallets/delegation/src/benchmarking.rs [hunk 10]
```diff
@@ -226,28 +238,35 @@ benchmarks! {
...
+		let delegate_acc_id: T::AccountId = delegate_acc_public.into();
+
 		let delegation_id = generate_delegation_id::<T>(u32::MAX);
-		let parent_id = leaf_id;
+		let parent_id = hierarchy_id;
 
 		let perm: Permissions = Permissions::ATTEST | Permissions::DELEGATE;
 		let hash_root = Pallet::<T>::calculate_delegation_creation_hash(&delegation_id, &hierarchy_id, &parent_id, &perm);
-		let sig = sp_io::crypto::sr25519_sign(KeyTypeId(*b"aura"), &delegate_acc_public, hash_root.as_ref()).ok_or("Error while building signature of delegation.")?;
+		let sig = (delegate_acc_id.clone(), AsRef::<[u8]>::as_ref(&hash_root).to_vec()).into();
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/delegation/src/benchmarking.rs:262]: I added a comment to explain what's happening. bdc6caffba84e0db510d6428bef72ff97e8ac239

File: pallets/delegation/src/tests.rs (hunk 1 above)
Comment (weichweich) [reply pallets/delegation/src/tests.rs:147]: 218f11688d3633c4d3af0737a1c1c1fbd3259df2

File: pallets/did/src/benchmarking.rs (hunk 7 above)
Comment (weichweich) [reply pallets/did/src/benchmarking.rs:770]: 6b058202dbd346633063c762ed0ef0f90a2898f1

File: pallets/did/src/benchmarking.rs (hunk 6 above)
Comment (weichweich) [reply pallets/did/src/benchmarking.rs:104]: 6b058202dbd346633063c762ed0ef0f90a2898f1

File: pallets/did/src/benchmarking.rs (hunk 8 above)
Comment (weichweich) [reply pallets/did/src/benchmarking.rs:797]: 6b058202dbd346633063c762ed0ef0f90a2898f1

File: support/src/signature.rs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+pub enum SignatureVerificationError {
+	/// The delegate's information is not present on chain.
+	SignerInformationNotPresent,
+	/// The signature over the delegation information is invalid.
+	SignatureInvalid,
+}
+
+/// Trait to implement to provide to the delegation pallet signature
+/// verification over a delegation details.
+pub trait VerifyDelegateSignature {
```
[the remark this answers is not in the document]
Comment (weichweich) [reply support/src/signature.rs:39]: We should keep it generic. 98cff4a1f200c8d53cbc4525a9fe1214a1d4f9ab

File: support/src/signature.rs (hunk 2 above)
Comment (weichweich) [reply support/src/signature.rs:86]: Yeah was a leftover from when I wanted to create a valid signature inside the trait. 94f13f93ed91eb8a7dde92509c6b71a2d195db1e

File: pallets/did/src/benchmarking.rs (hunk 6 above)
Comment (weichweich) [reply pallets/did/src/benchmarking.rs:104]: yeah didn't read this properly

**Review by wischli — APPROVED** [review]
LGTM
